### PR TITLE
prometheus: query-field: monaco: implement open-help-by-default

### PR DESCRIPTION
**Pull request by gabor**
in the prometheus query field, in the monaco-version, when an autocomplete window opens for the list of metrics for example, if you press ctrl+enter there, we open an additional help-window with extra help.
we want this extra-help-window to open by default, when the autocomplete-window opens.

this pull request does that.

there is unfortunately no simple way to do it, the way to do it is described in detail in the source code.

how to test:
- open the prometheus query editor
- start writing a metric-name, so that you get the metric-name-list autocomplete. it should by-defrault open the extra help window too

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    remove: (key: string, scope: unknown): void => {
+      strings.delete(key);
+    },
+
+    keys: (scope: unknown, target: unknown): string[] => {
+      return Array.from(strings.keys());
+    },
+
+    logStorage: (): void => {
+      console.log('storage log');
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:84]: Should we console.log here something else rather then just `storage log`? Not sure why would this be useful.

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      // we do not implement this
+      return Promise.resolve(undefined);
+    },
+  };
+}
+
+let overrideServices: monacoTypes.editor.IEditorOverrideServices | null = null;
+
+export function getOverrideServices(): monacoTypes.editor.IEditorOverrideServices {
+  // the idea is that we will only have one instance of this for every query editor
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:107]:
```suggestion
  // only have one instance of this for every query editor
```

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import { monacoTypes } from '@grafana/ui';
+
+// this thing here is a workaround in a way.
+// what we want to achieve, is that when the autocomplete-window
+// opens, the "second, extra popup" with the extra help,
+// also opens automatically.
+// but there is no API to achieve it.
+// the way to do it is to implement the `storageService`
+// interface, and provide our custom implementation,
+// which will defeault to `true` for the correct string-key.
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:10]:
```suggestion
// which will default to `true` for the correct string-key.
```

**Review by ivanahuckova — COMMENTED** [review]
Left some comments, but I have 1 UX question, should the "extra help" appear also when I hover over the metric? Because now it shows when I go up-down with arrows, but not when I hover over it.

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+// so we cannot rely on typescript to make sure
+// we do it right. all we can do is to manually
+// lookup the interface, and make sure we code our code right.
+// our code is a "best effort" approach,
+// i am not 100% how the `scope` and `target` things work,
+// but so far it seems to work ok.
+// i would use an another approach, if there was one available.
+
+function makeStorageService() {
+  // we need to return an object that fullfills this interface:
```
Comment (ifrost) [inline public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:22]:
```suggestion
  // we need to return an object that fulfills this interface:
```

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts (hunk 1 above)
Comment (gabor) [reply public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts]: i updated it to console-log this message: "logStorage: not implemented".

you see the problem is, i have this method only because the interface we need to implement wants this method to exist :-), so i have it. the documentation says it should log the storage to the console. ( https://github.com/microsoft/vscode/blob/ff1e16eebb93af79fd6d7af1356c4003a120c563/src/vs/platform/storage/common/storage.ts#L137 )

but i do not even know how to trigger this line of code, so i don't have any way to test it. i don't know how often this will be called, how fast it should be...

alternatively i could have an implementation like this:
```javascript
console.log(JSON.stringify(Array.from(strings.entries()))))
```

but i don't know if this would slow down something or not :-(

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    keys: (scope: unknown, target: unknown): string[] => {
+      return Array.from(strings.keys());
+    },
+
+    logStorage: (): void => {
+      console.log('storage log');
+    },
+
+    migrate: (): Promise<void> => {
+      // we do not implement this
```
Comment (ifrost) [inline public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:88]: What's the consequence of not implementing this (and methods below)? As in, what could possibly go wrong in the future? ;)

**Review by ifrost — COMMENTED** [review]
As there's no way to ensure we conform with Monaco API (no type checking) - would it be possible to add a small integration tests to ensure it works? This will help in case there's any breaking change in the API.

**Comment by gabor** [comment]
> Left some comments, but I have 1 UX question, should the "extra help" appear also when I hover over the metric? Because now it shows when I go up-down with arrows, but not when I hover over it.

unfortunately i don't think we can change this behavior :-( . it seems to be a monaco-thing. you can see it in visual-studio-code too, like this:
- trigger the autocomplete somehow, where it offers a list of things, for example by pressing ctrl+enter
- when the autocomplete popup shows, press ctrl+enter one more time, this opens the extra-help window
- now you can mouse-hover to other items, but the help will not get updated. even visually, it keeps the "item chosen by keyboard" highlighed with a different color, compared to the mouse-hover color

i hope this explains the situation.

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts (hunk 5 above)
Comment (gabor) [reply public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:88]: i have absolutely no idea :-) 

the only source of info we have is the comments in the interface-definition-file ( https://github.com/microsoft/vscode/blob/ff1e16eebb93af79fd6d7af1356c4003a120c563/src/vs/platform/storage/common/storage.ts#L37 ), and the fact that it does not crash when we use the monaco-query-editor.

for `migrate`:

i looked through the visual-studio-code, and all i find are instances like:
```typescript
	async migrate(toWorkspace: IWorkspaceInitializationPayload): Promise<void> {
		throw new Error('Migrating storage is currently unsupported in Web');
	}
```
and
```typescript
	async migrate(toWorkspace: IWorkspaceInitializationPayload): Promise<void> {
		// not supported
	}
```

so i THINK maybe it's ok to not implement it.

for `isNew`:

the interface-docs say it should return whether it was created in this "session" or not. so i changed the code to always return true, because we do not persist it. i'm not sure of course if this is the right way, but it's my best guess. i checked the usage in the vscode source code, and there is a storage-thing that simply stores it in a value and returns it, and it is set only from the electron-version of the code, based on some logic.

for `flush`:
the interface-docs say it is so that the storage can flush it's state, for example when a shutdown will happen. i don't want to persist anything, so i just went with not doing anything there.

in short, this is not the nicest code we have here, unfortunately there is no other way to achieve what we need (auto-open extra-help window) :-(

**Comment by gabor** [comment]
> As there's no way to ensure we conform with Monaco API (no type checking) - would it be possible to add a small integration tests to ensure it works? This will help in case there's any breaking change in the API.

good idea, i'll look into it. (basically we need to test that when the editor is shown, you should not get anything written to the error-console)

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts (hunk 5 above)
Comment (ifrost) [reply public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts:88]: Thanks for more info!

File: public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts (hunk 1 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts]: Yeah, makes sense to just log string. We don't have to do `console.log(JSON.stringify(Array.from(strings.entries()))))`. Thanks for explanation @gabor!

**Review by ifrost — APPROVED** [review]

**Comment by gabor** [comment]
> > As there's no way to ensure we conform with Monaco API (no type checking) - would it be possible to add a small integration tests to ensure it works? This will help in case there's any breaking change in the API.
> 
> good idea, i'll look into it. (basically we need to test that when the editor is shown, you should not get anything written to the error-console)

after discussion with @ifrost we decided we'll do the integration test in a separate PR later.
